kernel/mmu.c

```diff
@@ -574,6 +575,8 @@ void *k_mem_map(size_t size, uint32_t flags)
 	__ASSERT(page_frames_initialized, "%s called too early", __func__);
 	__ASSERT((flags & K_MEM_CACHE_MASK) == 0U,
 		 "%s does not support explicit cache settings", __func__);
+	__ASSERT(!size_add_overflow(size, CONFIG_MMU_PAGE_SIZE * 2, &total_size),
+			"too large size %zu passed to %s", size, __func__);
```

Comment:
If you expect to see this at runtime, it should be a runtime check and not an assert, no?  k_mem_map() is an app-visible API, not kernel-internal.  Asserts are for our bugs, not "theirs".

Also, given that this is an overflow condition, maybe a `MAX_MAP_SIZE` symbol with a comment explaining the guard pages would be easier to understand?  Or even a CONFIG_MAX_MAP_SIZE which apps or platforms could use to prevent ridiculously large mappings.

Reply:
This API is not accessible from userspace and other checks are done through asserts, kind of does not make sense only check this one. I can change it to use `CHECKIF`  but other checks are also valid and are being done using asserts ...

Reply:
I guess I don't understand the point then?  The commit message says "fix possible overflow", but the code only adds an __ASSERT.  By definition, it's not fixing an overflow.

Whether the other checks have value or not, I don't know.  Presumably they're there to check for regressions of previous bugs in the up-stack userspace code.  But regardless, you can't use __ASSERT to handle a runtime error, because it won't do it in production.

Reply:
We have asserts all over the code base to catch / point errors during development phase and we expect that once validation is done asserts are disabled in production. The exception is generally for input that comes from untrusted sources like i/o interfaces and USERSPACE. That is not the case for this API.

For the consistent sake, why the alignment check is not a runtime thing then ? Or the check for size == 0 ?

Reply:
> Asserts are for our bugs, not "theirs".

Generally I agree with that.

The line has maybe been blurred in the past for the sake of performance and code size, putting more of the onus on those deploying systems.

Reply:
Not denying that is a valid argument, but my point is that this is inconsistent with the rest of checks in this API and in others too.

Reply:
I'm going to leap into the future for a minute and say "CoPilot: What should we do?" 😅

Kidding of course.

But kind of in Andy's direction, I've recently started to convert some conditions I previously asserted to runtime checks.

The downside is measurable handful of .text bytes in tiny production memories and some thing else in a possibly hot path.

Here's an idea, 
```cpp
__ASSERT_OR(condition, message) {
  // incur a runtime check if condition fails
  // but can still evaluate to nothing based on condition
  return -42;
}
```

Reply:
@ceolin those other failures you're looking at are usage failures, they can't ever be right (i.e. using the API before initialization and passing a cache bit that isn't handled here).  They represent clear bugs.

An overflow is a **runtime** condition.  The addition almost always results in a valid address, but some apps, somewhere, in configurations we can't expect, might pass values that are too large and overflow.  There's no way we can say "That's Always Wrong" because we can't detect the app data from the future for lack of a time machine.

If we're concerned about overflow here, and I guess we are, we want this to be runtime.  Or otherwise to move the check up-stack into different APIs that are guaranteed to always be above us, etc...

Reply:
I've changed it, but I don't see how it is different from passing an unaligned size, it is also a runtime thing and it is based on the target's page size, just like the overflow.